api/envoy/config/listener/v3/listener_components.proto

```diff
@@ -193,6 +300,14 @@ message FilterChainMatch {
   //   and matching on values other than ``h2`` is going to lead to a lot of false negatives,
   //   unless all connecting clients are known to use ALPN.
   repeated string application_protocols = 10;
+
+  // Specifies the filter chain matching predicates to be used in the matching
+  // order. Each predicate extension must be specified at most once. For
+  // backwards compatibility, the existing fields in the filter chain match, if
+  // specified, are converted to their corresponding filter chain matching
+  // predicates in this list.
+  // [#not-implemented-hide:]
```

Comment:
I'm confused about the purpose of this field in relation to matching_order. Don't we basically just want a list of type extensions somewhere that will be the configurable matching order? Why do we need matching order above also?

Reply:
This is actually a set not a list of predicate extension configs for each predicate in the matching order. Not sure how else to express this concept. The order here is irrelevant.

Reply:
This is related to my comment above about matching order. I'm confused as to why we can't just have everything expressed in a single ordered list? Basically just force people to instantiate the list with the type urls in the order they want? The existing config can be translated to the ordered list?

Reply:
Each filter chain has different values for the predicates. Do you suggest each filter chain must have the same order for type URLs and then we deduce the global order from that? That makes sense.

Reply:
I think one annoyance is that empty predicates must be specified.

Reply:
OK I see the trade off, thanks for explaining. One one hand, it seems simpler if everything is embedded in the match and then we just verify that all of the chains have similar match ordering. On the other hand, there are downsides as you point out. I'm fine either way, but if we keep it like this can we beef up the docs? It's a bit confusing IMO.

Reply:
Yes, the docs are clearly lacking, and it's really confusing. I think we need to give proper examples and start from use cases. Let me try to get some istio examples, may be that will help clarifying the right API.